_posts/2021-07-13-ci-metrics.md

```diff
@@ -0,0 +1,116 @@
+
+But there is a wholly different version of that pain: When tests or the infrastructure slowly become worse. At first it's just a single "oh, a red test, this is unrelated, let's just retry", and before you know it it takes a day to get a pull request to green, and developers get frustrated and even afraid of touching tests.
+
+## Error budget
+
+With our tests and CI machinery being such important tools, we need to constantly measure their current state and quality to detect when they become too bad. In that case it is better to immediately react, stop feature development, and fix up our tools until they are in shape again. In total this minimizes the unnecessary pain and time spent on pull requests to get along with or work around breakage.
+
+The goal is to turn a vague gut feeling into objective numbers and alerts. This helps your own team to stay honest, be transparent to colleagues, users, and management, and also to invest the *right amount* of work into infrastructure. The latter point is crucial -- clearly you can spend too little work and end up with a tool which is more of a hindrance than an aid. But you can also spend *too much* work on it, just because it's fun, or you are a perfectionist, ending up with [diminishing returns](https://en.wikipedia.org/wiki/Diminishing_returns).
+
+This approach is called an *error budget* and descibed in all of its details and consequences in the free Google book [Site Reliability Engineering](https://sre.google/sre-book/table-of-contents/). This principle is mostly applied for services towards external customers, such as a web application or cloud engine. But we realized that it equally well applies to our development team as an internal service -- we are just our own customers, with two different hats: Most of the time we develop Cockpit, but sometimes we switch roles and develop our infrastructure.
+
+## Define Service Level Objectives
+
+The first step of that story is to sit down with the stakeholders -- in this case our own development team. We discussed our high-level expectations on our development process, and agreed that we are "happy" as long as these hold:
+
+ - PRs get validated in a reasonable time (queue + test run time).
+ - We don't waste time on interpreting unstable test results.
+ - We are not afraid of touching code.
+ - Test failures are relevant and meaningful. Relieve us from having to decide about “unrelated or not” every. single. time.
+
+These expectations are not yet objective; they need to be translated into concrete and measurable "Service Level Indicators" (SLI), together with a definition which range of an SLI we consider as "good"/"keeps us happy" according to the abstract expectations above. The latter is called a [Service Level Objective (SLO)](https://en.wikipedia.org/wiki/Service-level_objective). For us they look like this, in descending importance:
```

Comment:
"As stated, these expectations are not objective;"

Reply:
This is a grammar fix. Otherwise the "yet" seems to indicate that you've not yet done the follow on activity of defining metrics. But you have.

Reply:
Thanks! Done.